modules/packages/Python.chpl

```diff
@@ -0,0 +1,1417 @@
+       Do not create more than one instance of this class.
+  */
+  class Interpreter {
+    @chpldoc.nodoc
+    var converters: List.list(owned TypeConverter);
+
+    @chpldoc.nodoc
+    proc init() throws {
+      init this;
+
+      // preinit
+      var preconfig: PyPreConfig;
+      PyPreConfig_InitPythonConfig(c_ptrTo(preconfig));
+      preconfig.utf8_mode = 1;
+      checkPyStatus(Py_PreInitialize(c_ptrTo(preconfig)));
+
+      // init
+      var config_: PyConfig;
+      var cfgPtr = c_ptrTo(config_);
+      PyConfig_InitPythonConfig(cfgPtr);
+      defer PyConfig_Clear(cfgPtr);
+
+      // set program name
+      checkPyStatus(
+        PyConfig_SetString(
+          cfgPtr, c_ptrTo(config_.program_name), "chapel".c_wstr()));
+
+      // check VIRTUAL_ENV, if its set, make it the executable
+      var venv = getenv("VIRTUAL_ENV".c_str());
+      if venv != nil {
+        // ideally this just sets config_.home
+        // but by setting executable we can reuse the python logic to determine
+        // the locale (in the string sense, not the chapel sense)
+        const executable = string.createBorrowingBuffer(venv) + "/bin/python";
+        checkPyStatus(
+          PyConfig_SetString(
+            cfgPtr, c_ptrTo(config_.executable), executable.c_wstr()));
+      }
+
+      // initialize
+      checkPyStatus(Py_InitializeFromConfig(cfgPtr));
+
+      // add the current directory to the python path
+      var sys = PyImport_ImportModule("sys");
+      this.checkException();
+      var path = PyObject_GetAttrString(sys, "path");
+      this.checkException();
+      PyList_Insert(path, 0, Py_BuildValue("s", "."));
+      this.checkException();
+      Py_DECREF(path);
+      Py_DECREF(sys);
+    }
+    @chpldoc.nodoc
+    proc deinit() {
+      Py_Finalize();
+    }
+    /*
+      Run a string of python code.
+
+      This function will just run the code, it cannot be passed arguments or
+      return values.
+    */
+    proc run(code: string) throws {
+      PyRun_SimpleString(code.c_str());
+      this.checkException();
+    }
+
+    /*
+      Register a custom :type:`TypeConverter` with the interpreter. This allows
+      arbitrary Chapel types to be serialized/deserialized to and from Python
+      objects.
+
+      The registered converter will take precedence over the default converters,
+      allowing users to override the default behavior.
+    */
+    proc registerConverter(in cvt: owned TypeConverter) {
+      converters.pushBack(cvt);
+    }
+
+    /*
+      Compile a lambda function like 'lambda x,: x + 1'
+
+      Note: this only works with lambdas that a tuple of arguments, like 'x,' or '*args'
+    */
+    @chpldoc.nodoc
+    proc compileLambda(l: string): owned PyObject throws {
+      var globals = chpl_PyEval_GetFrameGlobals();
+      // create a globals if it doesn't exist
+      if !globals {
+        globals = Py_BuildValue("{}"); // TODO: need to decrement this?
```

Comment:
you do need to decref this. Normally, Python increfs automatically for things that return objects, including `GetFrameGlobals`, so I'd expect this to be safe. If not, set a bool flag if you use `Py_BuildValue`, and decref then.